### PR TITLE
[dualtor] Add script to verify consistency between kernel and ASIC 

### PR DESCRIPTION
**Pull request by lolyu**
#### What I did
Add script `dualtor_neighbor_check.py` to verify the neighbor consistency
based on the mux state. It will have the following output:
```
NEIGHBOR       MAC                PORT        MUX_STATE    IN_MUX_TOGGLE    NEIGHBOR_IN_ASIC    TUNNERL_IN_ASIC    HWSTATUS
-------------  -----------------  ----------  -----------  ---------------  ------------------  -----------------  ----------
192.168.0.2    ee:86:d8:46:7d:01  Ethernet4   standby      False            no                  yes                consistent
192.168.0.3    86:73:c2:22:bf:02  Ethernet8   standby      False            no                  yes                consistent
192.168.0.24   56:a6:bf:c5:dd:17  Ethernet92  active       False            yes                 no                 consistent
192.168.0.25   3a:18:56:f5:02:18  Ethernet96  active       False            yes                 no                 consistent
192.168.0.100  00:00:00:00:00:00  N/A         N/A          N/A              no                  yes                consistent
```

Signed-off-by: Longxiang Lyu <lolv@microsoft.com>

#### How I did it
the workflow of this scripts:
1. for non-zero-mac neighbors in `APPL_DB` `NEIGH_TABLE`, use the `ASIC_DB` fdb entries to find the mux port that it belongs to.
2. check if the neighbor is consistent with mux state:
    * if mux state is `active`, the neighbor is consistent only if the neighbor is present in `ASIC_DB` but no tunnel route.
    * if mux state is `standby`, the neighbor is consistent only if the tunnel route is present in `ASIC_DB`1 but no neighbor.
3. if there are any inconsistent neighbors and the mux port is currently in-toggle, the script will have a non-zero negative return, and will write error messages to logs.

#### How to verify it
UT and verify on testbed.

### CODE REVIEW

File: scripts/dualtor_neighbor_check.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+    WRITE_LOG_DEBUG("Command output: %s", output)
+    WRITE_LOG_DEBUG("Command return code: %s", p.returncode)
+    if p.returncode != 0:
+        raise RuntimeError("Command failed with return code %s: %s" % (p.returncode, output))
+    return output.decode()
+    
+
+def read_tables_from_db():
+    """Reads required tables from db."""
+    load_cmd = "sudo redis-cli SCRIPT LOAD \"%s\"" % DB_READ_SCRIPT
```
Comment (bingwang-ms) [inline scripts/dualtor_neighbor_check.py:291]: We are loading the same lua script at each run. Will it consume more and more memory? Or the previous one is overwriten?

File: scripts/dualtor_neighbor_check.py [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+
+def check_neighbor_consistency(neighbors, mux_states, hw_mux_states, mac_to_port_name_map,
+                               asic_route_table, asic_neigh_table, mux_server_to_port_map):
+    """Checks if neighbors are consistent with mux states."""
+
+    asic_route_destinations = set(json.loads(_)["dest"].split("/")[0] for _ in asic_route_table)
+    asic_neighs = set(json.loads(_)["ip"] for _ in asic_neigh_table)
+
+    check_results = []
+    show_items = ["NEIGHBOR", "MAC", "PORT", "MUX_STATE", "IN_MUX_TOGGLE", "NEIGHBOR_IN_ASIC", "TUNNRL_IN_ASIC", "HWSTATUS"]
```
Comment (bingwang-ms) [inline scripts/dualtor_neighbor_check.py:364]:
```suggestion
    show_items = ["NEIGHBOR", "MAC", "PORT", "MUX_STATE", "IN_MUX_TOGGLE", "NEIGHBOR_IN_ASIC", "TUNNEL_IN_ASIC", "HWSTATUS"]
```

**Review by Ndancejic — DISMISSED** [review]

File: scripts/dualtor_neighbor_check.py (hunk 1 above)
Comment (lolyu) [reply scripts/dualtor_neighbor_check.py]: Add support to cache the lua sha1 into a temporary key in `APPL_DB`.

**Comment by yxieca** [comment]
@lolyu Are you able to pass the coverage checker?

**Comment by bingwang-ms** [comment]
Discussed with Guohan offline, is it better if we integrate this script with existing route_checker.py? Both scripts are for consistency checking.

**Comment by lolyu** [comment]
> Discussed with Guohan offline, is it better if we integrate this script with existing route_checker.py? Both scripts are for consistency checking.

It is hard to integrate this with `route_checker.py`.
First, `dualtor_neighbor_check.py` has more tables to read, and it is based on a `Lua` script to read the db, and `route_check.py` uses the `swsscommon` table subscriptions to read table contents and its updates, its two different design.
Second, the mux states changes much frequently than the routes, especially in the mux flapping scenario, `route_check.py` reads the tables with `swsscommon` table subscriptions, so it is hard to capture a snapshot of the database and get accurate check result.

**Comment by lolyu** [comment]
Hi @bingwang-ms, @prsunny, could you please take a review, thanks!

File: scripts/dualtor_neighbor_check.py [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+    if_br_oid_to_port_name_map = {}
+    for if_br_oid, if_oid in if_br_oid_map.items():
+        if if_oid in port_name_map:
+            if_br_oid_to_port_name_map[if_br_oid] = port_name_map[if_oid]
+    return if_br_oid_to_port_name_map
+
+
+def is_dualtor(config_db):
+    device_metadata = config_db.get_table('DEVICE_METADATA')
+    subtype = device_metadata['localhost'].get('subtype', '')
```
Comment (bingwang-ms) [inline scripts/dualtor_neighbor_check.py:340]: Please handle `KeyError`.

Comment (lolyu) [reply scripts/dualtor_neighbor_check.py]: Done

**Review by bingwang-ms — DISMISSED** [review]

**Review by bingwang-ms — APPROVED** [review]

**Comment by lolyu** [comment]
Hi @yxieca, could you please help cherry-pick this into `202211` branch?

**Pull request merged**